lib/datasource/terraform-provider/index.ts

```diff
@@ -116,4 +123,114 @@ export class TerraformProviderDatasource extends TerraformDatasource {
     logger.trace({ dep }, 'dep');
     return dep;
   }
+
+  @cache({
+    namespace: `datasource-${TerraformProviderDatasource.id}-builds`,
+    key: (registryURL: string, repository: string, version: string) =>
+      `${registryURL}/${repository}/${version}`,
+  })
+  async getBuilds(
+    registryURL: string,
+    repository: string,
+    version: string
+  ): Promise<TerraformBuild[]> {
+    if (registryURL === TerraformProviderDatasource.defaultRegistryUrls[1]) {
+      // check if registryURL === secondary backend
+      const repositoryRegexResult =
+        TerraformProviderDatasource.repositoryRegex.exec(repository);
+      if (!repositoryRegexResult) {
+        // non hashicorp builds are not supported with releases.hashicorp.com
+        return null;
+      }
+      const lookupName = repositoryRegexResult.groups.lookupName;
+      const backendLookUpName = `terraform-provider-${lookupName}`;
+      let versionReleaseBackend: VersionDetailResponse;
+      try {
+        versionReleaseBackend = await this.getReleaseBackendIndex(
+          backendLookUpName,
+          version
+        );
+      } catch (err) {
+        logger.debug(
+          { err, backendLookUpName, version },
+          `Failed to retrieve builds for ${backendLookUpName} ${version}`
+        );
+        return null;
+      }
+      return versionReleaseBackend.builds;
+    }
+
+    // check public or private Terraform registry
+    const serviceDiscovery = await this.getTerraformServiceDiscoveryResult(
+      registryURL
+    );
+    if (!serviceDiscovery) {
+      logger.trace(`Failed to retrieve service discovery from ${registryURL}`);
+      return null;
+    }
+    const backendURL = `${registryURL}${serviceDiscovery['providers.v1']}${repository}`;
+    const versionsResponse = (
+      await this.http.getJson<TerraformRegistryVersions>(
+        `${backendURL}/versions`
+      )
+    ).body;
+    if (!versionsResponse.versions) {
+      logger.trace(`Failed to retrieve version list for ${backendURL}`);
+      return null;
+    }
+    const builds = versionsResponse.versions.find(
+      (value) => value.version === version
+    );
+    if (!builds) {
+      logger.trace(
+        `No builds found for ${repository}:${version} on ${registryURL}`
+      );
+      return null;
+    }
+    const result = await pMap(
+      builds.platforms,
+      async (platform) => {
+        const buildURL = `${backendURL}/${version}/download/${platform.os}/${platform.arch}`;
+        try {
+          const res = (
+            await this.http.getJson<TerraformRegistryBuildResponse>(buildURL)
+          ).body;
+          const newBuild: TerraformBuild = {
+            name: repository,
+            url: res.download_url,
+            version,
+            ...res,
+          };
+          return newBuild;
+        } catch (e) {
+          logger.debug({ error: e, url: buildURL }, 'Failed to retrieve build');
```

Comment:
```suggestion
        } catch (err) {
          logger.debug({ err, url: buildURL }, 'Failed to retrieve build');
```

simplify, we've a special serializer for `err` metadata logging.